### PR TITLE
Dashboards: Make path to default dashboard configurable

### LINKED ISSUE #25463: Make default home dashboard configurable in grafana.ini
It should be possible to switch the [path to the default home dashboard](https://github.com/grafana/grafana/blob/master/pkg/api/dashboard.go#L335) in `grafana.ini`.

This will make it possible to change the first impression using config files.

### PR DESCRIPTION
**Pull request by bergquist**
closes https://github.com/grafana/grafana/issues/25463

### CODE REVIEW

File: pkg/api/dashboard_test.go [hunk 1]
```diff
@@ -13,10 +14,58 @@ import (
...
+
+	for _, tc := range tcs {
+		dash := dtos.DashboardFullWithMeta{}
+		dash.Meta.IsHome = true
+		dash.Meta.CanEdit = req.SignedInUser.HasRole(models.ROLE_EDITOR)
+		dash.Meta.FolderTitle = "General"
+
+		homeDashJSON, err := ioutil.ReadFile(tc.expectedDashboardPath)
+		setting.DefaultHomeDashboardPath = tc.defaultSetting
+		require.NoError(t, err)
```
Comment (aknuds1) [inline pkg/api/dashboard_test.go:53]: Can you move this line to 52, so it's co-located with the originating statement?

File: pkg/setting/setting.go [hunk 2]
```diff
@@ -272,6 +273,11 @@ type Cfg struct {
 	DisableSanitizeHtml              bool
 	EnterpriseLicensePath            string
 
+	// Dashboards
+	DashboardVersionsToKeep  int
+	MinRefreshInterval       string
```
Comment (aknuds1) [inline pkg/setting/setting.go:278]: Why is it necessary to add also these two fields as part of this PR?

File: pkg/setting/setting.go [hunk 3]
```diff
@@ -776,6 +782,15 @@ func (cfg *Cfg) Load(args *CommandLineArgs) error {
 		return err
 	}
 
+	DefaultHomeDashboardPath = dashboards.Key("default_home_dashboard_path").MustString("")
+	cfg.DefaultHomeDashboardPath = DefaultHomeDashboardPath
+	if err != nil {
+		return err
+	}
```
Comment (aknuds1) [inline pkg/setting/setting.go:789]: I think this must be left over?? I don't see any connected origin of `err`.
```suggestion
```

File: pkg/setting/setting.go [hunk 4]
```diff
@@ -776,6 +782,15 @@ func (cfg *Cfg) Load(args *CommandLineArgs) error {
...
+
+	cfg.DashboardVersionsToKeep = DashboardVersionsToKeep
+	cfg.MinRefreshInterval = MinRefreshInterval
```
Comment (aknuds1) [inline pkg/setting/setting.go:792]: Why must these be added for this PR?

**Review by aknuds1 — COMMENTED** [review]
Please see my questions.

File: pkg/setting/setting.go (hunk 4 above)
Comment (bergquist) [reply pkg/setting/setting.go:789]: Good catch I'll address that

File: pkg/api/dashboard.go [hunk 5]
```diff
@@ -315,15 +314,16 @@ func dashboardSaveErrorToApiResponse(err error) Response {
 	return Error(500, "Failed to save dashboard", err)
 }
 
-func GetHomeDashboard(c *models.ReqContext) Response {
+// GetHomeDashboard return the home dashboard
```
Comment (aknuds1) [inline pkg/api/dashboard.go:317]:
```suggestion
// GetHomeDashboard returns the home dashboard.
```

File: pkg/api/dashboard_test.go [hunk 6]
```diff
@@ -26,6 +26,10 @@ func TestGetHomeDashboard(t *testing.T) {
 		HomePath: "../../",
 	})
 
+	hs := &HTTPServer{
+		Cfg: cfg,
+	}
+
```
Comment (aknuds1) [inline pkg/api/dashboard_test.go:32]: Can you move this until after the error check? It's confusing now that the error check is located away from line 25.

File: pkg/api/dashboard.go [hunk 7]
```diff
@@ -315,15 +314,16 @@ func dashboardSaveErrorToApiResponse(err error) Response {
...
+func (hs *HTTPServer) GetHomeDashboard(c *models.ReqContext) Response {
 	prefsQuery := models.GetPreferencesWithDefaultsQuery{User: c.SignedInUser}
-	if err := bus.Dispatch(&prefsQuery); err != nil {
+	if err := hs.Bus.Dispatch(&prefsQuery); err != nil {
```
Comment (aknuds1) [inline pkg/api/dashboard.go:320]: Is it part of some general improvements to use a per-object bus instance instead? I don't know the ramifications here.

**Review by aknuds1 — COMMENTED** [review]
Fixed a typo and left some questions.

File: pkg/api/dashboard.go (hunk 7 above)
Comment (bergquist) [reply pkg/api/dashboard.go:320]: Yes. Same as with the `settings` package. But it doesn't look like we mention it in the backend readme. 

https://github.com/grafana/grafana/blob/master/pkg/README.md#settings-refactoring

File: docs/sources/installation/configuration.md [hunk 8]
```diff
@@ -525,6 +525,11 @@ Number dashboard versions to keep (per dashboard). Default: `20`, Minimum: `1`.
 This will restrict users to set the refresh interval of a dashboard lower than given interval. Per default this is 5 seconds.
 The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. `30s` or `1m`.
 
+### default_home_dashboard_path
+
+Path to the default home dashboard. If this value is empty, Grafana will default to use StaticRootPath + "dashboards/home.json"
```
Comment (oddlittlebird) [inline docs/sources/installation/configuration.md:530]:
```suggestion
Path to the default home dashboard. If this value is empty, then Grafana uses StaticRootPath + "dashboards/home.json"
```

**Review by aknuds1 — APPROVED** [review]
LGTM

File: conf/sample.ini [hunk 9]
```diff
@@ -233,6 +233,9 @@
 # The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
 ;min_refresh_interval = 5s
 
+# Path to the default home dashboard. If this value is empty, Grafana will default to use StaticRootPath + "dashboards/home.json"
```
Comment (bergquist) [inline conf/sample.ini:236]:
```suggestion
# Path to the default home dashboard. If this value is empty, then Grafana uses StaticRootPath + "dashboards/home.json"
```

File: conf/defaults.ini [hunk 10]
```diff
@@ -234,6 +234,9 @@ versions_to_keep = 20
 # The interval string is a possibly signed sequence of decimal numbers, followed by a unit suffix (ms, s, m, h, d), e.g. 30s or 1m.
 min_refresh_interval = 5s
 
+# Path to the default home dashboard. If this value is empty, Grafana will default to use StaticRootPath + "dashboards/home.json"
```
Comment (bergquist) [inline conf/defaults.ini:237]:
```suggestion
# Path to the default home dashboard. If this value is empty, then Grafana uses StaticRootPath + "dashboards/home.json"
```

**Review by marefr — APPROVED** [review]
LGTM

**Comment by geekq** [comment]
Very nice feature, thanks! I've found the documentation a bit confusing though!

What is "Path" supposed to be, url path, file path, relative or absolute? What is StaticRootPath?
It came out, it is an absolute file path inside container where Grafana server is running.

By changing

    # Path to the default home dashboard. If this ...

to

    # Absolute file path to the default home dashboard. If this ...

some support effort can be possibly saved.

**Comment by FireEmerald** [comment]
Had the same confusion as **geekq** about the `StaticRootPath`. The documentation should be improved.

```
$ docker exec -it grafana_container_name bash
bash-5.0$ ls
LICENSE          VERSION          plugins-bundled
NOTICE.md        bin              public
README.md        conf             scripts

bash-5.0$ cd public/ <-- "static_root_path = public" from config

bash-5.0$ ls
app         dashboards  fonts       robots.txt  test        views
build       emails      img         sass        vendor

bash-5.0$ cd dashboards/

bash-5.0$ ls
default.json           scripted.js            scripted_templated.js
home.json <-- Default home dashboard
```

**Pull request merged**